src/Match.hs

```diff
@@ -0,0 +1,94 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Match (on, matche, example') where
+
+import Core
+import UnifiableBase ()
+import Goal
+import Control.Lens ( Prism', review, Prism, makePrisms )
+import Data.Void (Void)
+import Data.Function ((&))
+
+on :: (Unifiable s, Fresh a)
+   => Prism' (Term s) a
+   -> (a -> Goal x)
+   -> (ValueOrVar s -> Goal x)
+   -> ValueOrVar s
+   -> Goal x
+on p f g x = disj (g x) thisArm
+  where
+    thisArm = case x of
+      Var _ -> fresh $ \vars -> do
+        x === Value (review p vars)
+        f vars
+      Value value -> case p Left value of
+        Left a -> f a
+        Right _ -> failo
+
+matche :: ValueOrVar a -> Goal x
+matche = const failo
+
+---
+
+data LogicEither a b
+  = LogicLeft (ValueOrVar a)
+  | LogicRight (ValueOrVar b)
+makePrisms ''LogicEither
+
+instance Unifiable Void where
+  type Term Void = Void
+
+instance (Unifiable a, Unifiable b) => Unifiable (Either a b) where
+  type Term (Either a b) = LogicEither a b
+
+  subst f (LogicLeft x) = LogicLeft (subst' f x)
+  subst f (LogicRight y) = LogicRight (subst' f y)
+
+  inject (Left x) = LogicLeft (inject' x)
+  inject (Right y) = LogicRight (inject' y)
+
+  extract (LogicLeft x) = Left <$> extract' x
+  extract (LogicRight x) = Right <$> extract' x
+
+  unify (LogicLeft x) (LogicLeft y) state = unify' x y state
+  unify (LogicRight x) (LogicRight y) state = unify' x y state
+  unify _ _ _ = Nothing
+
+-- example :: ValueOrVar (Maybe Int) -> Goal ()
+-- example = matche
+--   & on _LogicNothing return
+--   & on _LogicJust (\x -> x === 0)
+
+example' :: ValueOrVar (Either Bool Int) -> Goal ()
+example' = matche'
+  & on' _LogicLeft _LogicLeft (\x -> x === Value True)
+  & on' _LogicRight _LogicRight (\x -> x === Value 0)
+
+class Empty a
+
+instance Empty Void
+instance (Empty a, Empty b) => Empty (Either a b)
+
+matche' :: Empty a => ValueOrVar a -> Goal x
+matche' = const failo
+
+on' :: (Unifiable s, Fresh a)
+    => Prism' (Term s) a
+    -> Prism (Term s) (Term t) a (ValueOrVar Void)
```

Comment:
I don't really like that we have to take the prism twice, and I haven't found a way to use the same prism in both cases